File: Crittografia/crittografia.cpp
#include <bits/stdc++.h>

using namespace std;

bool is_int(char c){
	return (int(c) <= 57 && int(c) >= 48);
}

int main(){
	
	#ifdef EVAL
	freopen("input.txt", "r", stdin);
	freopen("output.txt", "w", stdout);
	#endif
	
	int N, L;
	
	cin >> N >> L;
	
	string S;
	
	cin >> S;
	
	vector<string> password(N, "nope");
	
	for (int i = 0; i < N; i++){
		password[i] = "nope";
	}
	
	bool done = false;
	
	string empty_char = "nope";
	
	for (int i = 0; i < int(L / 2); i++){
		
//		cout << i << "\n";
		
		if (is_int(S[i]) && !is_int(S[L - i - 1])){
			if (password[int(S[i]) - 48].size() > 1 || password[int(S[i]) - 48][0] == S[L - i - 1]){
				password[int(S[i]) - 48] = S[L - i - 1];
			} else {
				done = true;
				cout << "impossibile\n";
				break;
			}
		} else if (!is_int(S[i]) && is_int(S[L - i - 1])){
//			cout << "sd\n";
//			cout << password[int(S[L - i - 1]) - 48] << "\n";
			//cout << "|" << int(password[S[L - i - 1]]) << "|" << int(empty_char[0]) << "| --- " << S[i] << " a " << S[L - i - 1] << "\n";
//			cout << "yess\n";
			if (password[int(S[L - i - 1]) - 48].size() > 1 || password[int(S[L - i - 1]) - 48][0] == S[i]){
				password[int(S[L - i - 1]) - 48] = S[i];
			} else {
				done = true;
				cout << "impossibile\n";
				break;
			}
		}
		
	}
	
	if (!done){
		
		string result = "";
		for (int i = 0; i < N; i++){
			result += password[i];
		}
		
		if (result.size() != N){
			cout << "impossibile\n";
		} else {
			cout << result << "\n";
		}
		
	}
	
	
	return 0;
}

File: Crittografia/crittografia.py
#Scrivi qui il tuo codice
from __future__ import print_function
try: input = raw_input
except: pass

import sys 
sys.stdout = open("output.txt", "w")
sys.stdin = open("input.txt", "r")



N, L = map(int, input().strip().split())

S = str(input().strip())
S = [char for char in S]

for i in range(len(S)):
    try:
        S[i] = int(S[i])
    except:
        S[i] = str(S[i])


password = [''] * N


for i in range(int(L / 2)):
    #print(S[i], S[L - i - 1], type(S[i]))
    if type(S[i]) == int and type(S[L - i - 1]) == str:
        if password[S[i]] == '' or password[S[i]] == S[L - i - 1]:
            password[S[i]] = S[L - i - 1]
            #print("yuuu", password[S[i]])
        else:
            print("impossibile")
            break
    elif type(S[i]) == str and type(S[L - i - 1]) == int:
        if password[S[L - i - 1]] == '' or password[S[L - i - 1]] == S[i]:
            password[S[L - i - 1]] = S[i]
            #print("yeee", password[S[L - i - 1]])
        else:
            print("impossibile")
            break
else:
    pass
    result = "".join(password)
    
    if len(result) != N:
        #print(result)
        #print(password)
        print("impossibile")
    else:
        print(result)

